File: py_package/everyai_core/providers/cloudflare.py
"""Cloudflare AI provider implementation.

Queries Cloudflare's direct AI Run REST endpoints.
"""

import os
import json
from typing import Any, Generator
import httpx
from everyai_core.providers.base import BaseProvider
from everyai_core.exceptions import raise_for_status, ConfigurationError
from everyai_core.types import ChatCompletionResponse, ChatCompletionChoice, UsageInfo, ModelInfo


class CloudflareProvider(BaseProvider):
    """Provider class for Cloudflare AI API."""

    def chat(
        self,
        model: str,
        messages: list[dict[str, Any]],
        temperature: float = 0.7,
        max_tokens: int | None = None,
        stream: bool = False,
        **kwargs
    ) -> ChatCompletionResponse | Generator[ChatCompletionResponse, None, None]:
        """Send a chat completion request to Cloudflare direct AI Run endpoint."""
        # Resolve Account ID
        account_id = kwargs.pop("account_id", None) or self.extra_config.get("account_id") or os.environ.get("CLOUDFLARE_ACCOUNT_ID")
        if not account_id:
            raise ConfigurationError(
                "Cloudflare AI requires 'account_id'. Please pass it via "
                "EveryAI(provider_config={'cloudflare': {'account_id': 'your_id'}}), "
                "or set the environment variable 'CLOUDFLARE_ACCOUNT_ID'."
            )

        base = (self.base_url or f"https://api.cloudflare.com/client/v4/accounts/{account_id}/ai/run").rstrip("/")
        # Model identifier needs to be stripped/passed directly
        endpoint = f"{base}/{model.lstrip('/')}"

        headers = {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json",
        }

        # Build payload for Cloudflare AI Run Chat
        payload = {
            "messages": messages,
            "stream": stream,
        }
        # Add temperature if supported by payload schema
        if temperature != 0.7:
            payload["temperature"] = temperature
        if max_tokens is not None:
            payload["max_tokens"] = max_tokens
        
        # Merge other parameter overrides
        payload.update(kwargs)

        timeout = kwargs.pop("timeout", self.extra_config.get("timeout", 20.0))

        if not stream:
            response = httpx.post(endpoint, headers=headers, json=payload, timeout=timeout)
            if response.status_code != 200:
                raise_for_status("cloudflare", response.status_code, response.text)
            
            data = response.json()
            if not data.get("success"):
                errors = data.get("errors", [])
                err_msg = errors[0].get("message") if (errors and isinstance(errors[0], dict)) else "Cloudflare query failed"
                raise_for_status("cloudflare", 400, err_msg)

            result = data.get("result", {})
            content = result.get("response", "")

            choices = [
                ChatCompletionChoice(
                    index=0,
                    message={"role": "assistant", "content": content},
                    finish_reason="stop"
                )
            ]

            return ChatCompletionResponse(
                id=None,
                object="chat.completion",
                created=0,
                model=model,
                choices=choices,
                usage=UsageInfo(None, None, None),
                provider="cloudflare"
            )

        # Streaming execution
        def stream_generator() -> Generator[ChatCompletionResponse, None, None]:
            with httpx.Client() as client:
                try:
                    with client.stream("POST", endpoint, headers=headers, json=payload, timeout=timeout) as r:
                        if r.status_code != 200:
                            r.read()
                            raise_for_status("cloudflare", r.status_code, r.text)

                        for line in r.iter_lines():
                            if isinstance(line, bytes):
                                line = line.decode("utf-8")
                            line = line.strip()
                            if not line:
                                continue
                            if line.startswith("data: "):
                                data_str = line[len("data: "):]
                                if data_str == "[DONE]":
                                    break
                                try:
                                    chunk_data = json.loads(data_str)
                                    yield self._parse_chunk(chunk_data, model)
                                except Exception:
                                    continue
                except Exception as e:
                    raise e

        return stream_generator()

    def _parse_chunk(self, chunk_data: dict[str, Any], model: str) -> ChatCompletionResponse:
        """Parse Cloudflare SSE JSON chunk into a ChatCompletionResponse."""
        content = chunk_data.get("response", "")
        choices = [
            ChatCompletionChoice(
                index=0,
                message={"role": "assistant", "content": content},
                finish_reason=None
            )
        ]

        return ChatCompletionResponse(
            id=None,
            object="chat.completion.chunk",
            created=0,
            model=model,
            choices=choices,
            usage=None,
            provider="cloudflare"
        )

    def list_models(self) -> list[ModelInfo]:
        """List models available for Cloudflare AI."""
        # Cloudflare has standard model presets since listing requires pagination search APIs
        presets = [
            "@cf/meta/llama-3-8b-instruct",
            "@cf/meta/llama-3-70b-instruct",
            "@cf/mistral/mistral-7b-instruct-v0.1",
            "@cf/meta/llama-2-7b-chat-int8"
        ]
        return [ModelInfo(id=p, name=p.split("/")[-1], context_length=2048, owned_by="cloudflare") for p in presets]

File: js_package/everyai-core/src/providers/cloudflare.js
import { BaseProvider } from "./base.js";
import { raiseForStatus, ConfigurationError } from "../exceptions.js";
import { ChatCompletionResponse, ChatCompletionChoice, UsageInfo, ModelInfo } from "../types.js";

/**
 * Provider class for Cloudflare AI API.
 */
export class CloudflareProvider extends BaseProvider {
    /**
     * Send a chat completion request to Cloudflare direct AI Run endpoint.
     * 
     * @param {string} model The ID/name of the model to query.
     * @param {Array<Object>} messages A list of messages.
     * @param {number} [temperature] Sampling temperature (default: 0.7).
     * @param {number|null} [maxTokens] Maximum number of tokens to generate.
     * @param {boolean} [stream] Whether to stream the response.
     * @param {Object} [kwargs] Provider-specific overrides.
     * @returns {Promise<ChatCompletionResponse|AsyncGenerator<ChatCompletionResponse, void, unknown>>}
     */
    async chat(model, messages, temperature = 0.7, maxTokens = null, stream = false, kwargs = {}) {
        // Resolve Account ID
        const accountId = kwargs.accountId || kwargs.account_id || 
                          this.extraConfig.accountId || this.extraConfig.account_id || 
                          process.env.CLOUDFLARE_ACCOUNT_ID;
        if (!accountId) {
            throw new ConfigurationError(
                "Cloudflare AI requires 'account_id'. Please pass it via " +
                "new EveryAI({ providerConfig: { 'cloudflare': { 'account_id': 'your_id' } } }), " +
                "or set the environment variable 'CLOUDFLARE_ACCOUNT_ID'."
            );
        }

        const base = (this.baseUrl || `https://api.cloudflare.com/client/v4/accounts/${accountId}/ai/run`).replace(/\/$/, "");
        const cleanModel = model.replace(/^\//, "");
        const endpoint = `${base}/${cleanModel}`;

        const headers = {
            "Authorization": `Bearer ${this.apiKey}`,
            "Content-Type": "application/json",
        };

        const payload = {
            messages: messages,
            stream: stream,
        };
        if (temperature !== 0.7) {
            payload["temperature"] = temperature;
        }
        if (maxTokens !== null) {
            payload["max_tokens"] = maxTokens;
        }

        // Merge overrides
        Object.assign(payload, kwargs);

        const timeout = kwargs.timeout ?? this.extraConfig.timeout ?? 20.0;
        const controller = new AbortController();
        const timeoutId = setTimeout(() => controller.abort(), timeout * 1000);

        try {
            const response = await fetch(endpoint, {
                method: "POST",
                headers: headers,
                body: JSON.stringify(payload),
                signal: controller.signal
            });
            clearTimeout(timeoutId);

            if (response.status !== 200) {
                const text = await response.text();
                raiseForStatus("cloudflare", response.status, text);
            }

            if (!stream) {
                const data = await response.json();
                if (!data.success) {
                    const errors = data.errors || [];
                    const errMsg = (errors[0] && errors[0].message) ? errors[0].message : "Cloudflare query failed";
                    raiseForStatus("cloudflare", 400, errMsg);
                }

                const result = data.result || {};
                const content = result.response || "";

                const choices = [
                    new ChatCompletionChoice(
                        0,
                        { role: "assistant", content: content },
                        "stop"
                    )
                ];

                return new ChatCompletionResponse(
                    null,
                    "chat.completion",
                    0,
                    model,
                    choices,
                    new UsageInfo(null, null, null),
                    "cloudflare"
                );
            }

            // Streaming execution
            const self = this;
            async function* streamGenerator() {
                const reader = response.body.getReader();
                const decoder = new TextDecoder("utf-8");
                let buffer = "";

                try {
                    while (true) {
                        const { value, done } = await reader.read();
                        if (done) break;

                        buffer += decoder.decode(value, { stream: true });
                        const lines = buffer.split("\n");
                        buffer = lines.pop(); // keep trailing partial line

                        for (let line of lines) {
                            line = line.trim();
                            if (!line) continue;
                            if (line.startsWith("data: ")) {
                                const dataStr = line.slice(6);
                                if (dataStr === "[DONE]") {
                                    return;
                                }
                                try {
                                    const chunkData = JSON.parse(dataStr);
                                    yield self._parseChunk(chunkData, model);
                                } catch (e) {
                                    continue;
                                }
                            }
                        }
                    }
                } finally {
                    reader.releaseLock();
                }
            }

            return streamGenerator();
        } catch (error) {
            clearTimeout(timeoutId);
            throw error;
        }
    }

    /**
     * Parse Cloudflare streaming chunk into a ChatCompletionResponse.
     * 
     * @private
     * @param {Object} chunkData
     * @param {string} model
     * @returns {ChatCompletionResponse}
     */
    _parseChunk(chunkData, model) {
        const content = chunkData.response || "";
        const choices = [
            new ChatCompletionChoice(
                0,
                { role: "assistant", content: content },
                null
            )
        ];

        return new ChatCompletionResponse(
            null,
            "chat.completion.chunk",
            0,
            model,
            choices,
            null,
            "cloudflare"
        );
    }

    /**
     * List models available for Cloudflare AI (presets list).
     * 
     * @returns {Promise<Array<ModelInfo>>} A list of model info objects.
     */
    async listModels() {
        const presets = [
            "@cf/meta/llama-3-8b-instruct",
            "@cf/meta/llama-3-70b-instruct",
            "@cf/mistral/mistral-7b-instruct-v0.1",
            "@cf/meta/llama-2-7b-chat-int8"
        ];
        return presets.map(p => new ModelInfo(
            p,
            p.split("/").pop(),
            2048,
            "cloudflare"
        ));
    }
}
